File: src/stages/stageTwoVCP.js
function findPeaksAndTroughs(data, window = 3) {
  const peaks = [];
  const troughs = [];
  for (let i = window; i < data.length - window; i++) {
    const curHigh = data[i].high;
    const curLow = data[i].low;
    let isPeak = true;
    let isTrough = true;
    for (let j = 1; j <= window; j++) {
      if (data[i - j].high >= curHigh || data[i + j].high >= curHigh) {
        isPeak = false;
      }
      if (data[i - j].low <= curLow || data[i + j].low <= curLow) {
        isTrough = false;
      }
      if (!isPeak && !isTrough) break;
    }
    if (isPeak) {
      peaks.push({ index: i, price: curHigh, date: data[i].date });
    }
    if (isTrough) {
      troughs.push({ index: i, price: curLow, date: data[i].date });
    }
  }
  return { peaks, troughs };
}

function analyzeContractions(peaks, troughs) {
  const points = [
    ...peaks.map(p => ({ ...p, type: 'peak' })),
    ...troughs.map(t => ({ ...t, type: 'trough' }))
  ].sort((a, b) => a.index - b.index);

  const contractions = [];
  for (let i = 0; i < points.length - 1; i++) {
    const first = points[i];
    const second = points[i + 1];
    if (first.type === 'peak' && second.type === 'trough') {
      const depth = (first.price - second.price) / first.price;
      const duration = second.index - first.index;
      contractions.push({ depth, duration, peak: first, trough: second });
    }
  }
  return contractions;
}

function isVCP(contractions) {
  if (contractions.length < 2) return false;
  for (let i = 0; i < contractions.length - 1; i++) {
    if (!(contractions[i].depth > contractions[i + 1].depth)) {
      return false;
    }
  }
  return true;
}

function stageTwoVCP(data) {
  if (!Array.isArray(data) || data.length < 30) {
    return false;
  }

  const recentData = data.slice(-80);
  const { peaks, troughs } = findPeaksAndTroughs(recentData, 3);
  const contractions = analyzeContractions(peaks, troughs);
  return isVCP(contractions);
}

module.exports = stageTwoVCP;
module.exports.findPeaksAndTroughs = findPeaksAndTroughs;
module.exports.analyzeContractions = analyzeContractions;
module.exports.isVCP = isVCP;

File: src/stages/stage_two_vcp.py
from typing import List, Dict


def find_peaks_and_troughs(data: List[Dict], window: int = 3):
    peaks = []
    troughs = []
    for i in range(window, len(data) - window):
        cur_high = data[i]['high']
        cur_low = data[i]['low']
        is_peak = True
        is_trough = True
        for j in range(1, window + 1):
            if data[i - j]['high'] >= cur_high or data[i + j]['high'] >= cur_high:
                is_peak = False
            if data[i - j]['low'] <= cur_low or data[i + j]['low'] <= cur_low:
                is_trough = False
            if not is_peak and not is_trough:
                break
        if is_peak:
            peaks.append({'index': i, 'price': cur_high, 'date': data[i]['date']})
        if is_trough:
            troughs.append({'index': i, 'price': cur_low, 'date': data[i]['date']})
    return {'peaks': peaks, 'troughs': troughs}


def analyze_contractions(peaks, troughs):
    points = [
        *({**p, 'type': 'peak'} for p in peaks),
        *({**t, 'type': 'trough'} for t in troughs),
    ]
    points.sort(key=lambda x: x['index'])

    contractions = []
    for first, second in zip(points, points[1:]):
        if first['type'] == 'peak' and second['type'] == 'trough':
            depth = (first['price'] - second['price']) / first['price']
            duration = second['index'] - first['index']
            contractions.append({'depth': depth, 'duration': duration, 'peak': first, 'trough': second})
    return contractions


def is_vcp(contractions) -> bool:
    if len(contractions) < 2:
        return False
    for first, second in zip(contractions, contractions[1:]):
        if not (first['depth'] > second['depth']):
            return False
    return True


def stage_two_vcp(data: List[Dict]) -> bool:
    if len(data) < 30:
        return False
    recent = data[-80:]
    res = find_peaks_and_troughs(recent, 3)
    contractions = analyze_contractions(res['peaks'], res['troughs'])
    return is_vcp(contractions)
